src/mongocxx/test/client_session.cpp

```diff
@@ -60,6 +64,35 @@ TEST_CASE("session options", "[session]") {
         auto s = c.start_session(opts);
         REQUIRE(!s.options().causal_consistency());
     }
+
+    SECTION("set snapshot consistency") {
+        options::client_session opts;
+
+        REQUIRE_FALSE(opts.snapshot());
+        opts.causal_consistency(false);
+        opts.snapshot(true);
+        REQUIRE(opts.snapshot());
+        REQUIRE_FALSE(opts.causal_consistency());
+
+        auto s = c.start_session(opts);
+
+        REQUIRE(opts.snapshot());
+        REQUIRE_FALSE(opts.causal_consistency());
+    }
+
+    SECTION("causal and snapshot consistency are mutually exclusive") {
+        // Trying to set both casusal and snapshot consistency should result
+        // in an error when starting the session:
+        options::client_session opts;
+
+        opts.snapshot(true);
+        opts.causal_consistency(true);
+
+        REQUIRE(opts.causal_consistency());
+        REQUIRE(opts.snapshot());
+
+        REQUIRE_THROWS_AS(c.start_session(opts), mongocxx::exception);
+    }
```

Comment:
I'm fairly into implementing this... however two situations are occuring to me:

1) that the only case we actually care about, though, is when they're both actually set and contain a value of true. 

As I read it, anything else is not an error (see the truth table that you wrote out). The other conditions to check are that they are both "unset" (i.e. contain no value) at creation on the client side (e.g. one, another, or both are not actually specified), and then "set" (i.e. contain values-- we can't actually check the default on the client side because according to the spec the server may change it, at least in the case of causal_consistency), yet still mutually-exclusive. So this test may check... too much.

2) I'm not sure we can actually test this using this strategy:

Please note that I'm new to reading MongoDB spec-ese, so it's entirely possible that I'm misinterpreting the language. Here's what I'm reading from the Causal Consistency spec:

"The default value of this property is not supplied. If no value is supplied for causalConsistency the value will be inherited. Currently it is inherited from the global default which is defined to be true. In the future it might be inherited from client settings."

...we can't control the server's global default, and although IF we check it it should be true after being set by its default, we can only detect this change as a regression on the client. It may not be an actual error, we're checking server rather than client behavior. 

This is a bit of a head-scratcher. I'm going to err on the side of implementing a test like this, as the answer is probably that I'm misinterpreting the situation-- but when there's some time this is probably worth pondering. Thank you for highlighting this!

Reply:
Ok-- as per offline conversations and some more reading (and further head-scratching!), I've implemented a test that generates the permutations and expects failure from any condition where both of these were set to "true". The case of "what /would/ the server set these to" is implicitly included in that, and I have also added some further explicit tests; some of these are redundant but explicit.